fast_pauli/cpp/include/__summed_pauli_op.hpp

```diff
@@ -389,25 +563,40 @@ template <std::floating_point T> struct SummedPauliOp
         std::vector<std::complex<T>> expectation_vals_raw(n_pauli_strings() * n_data);
         Tensor<2> expectation_vals(expectation_vals_raw.data(), n_pauli_strings(), n_data);
 
-        if constexpr (std::is_same_v<ExecutionPolicy, std::execution::parallel_policy>)
+        if constexpr (is_parallel_execution_policy_v<ExecutionPolicy>)
         {
 
-#pragma omp parallel for schedule(static)
-            for (size_t j = 0; j < n_pauli_strings(); ++j)
+#pragma omp parallel
             {
-                std::mdspan expectation_vals_j = std::submdspan(expectation_vals, j, std::full_extent);
-                pauli_strings[j].expectation_value(expectation_vals_j, states);
-            }
+#pragma omp for schedule(static)
+                for (size_t j = 0; j < n_pauli_strings(); ++j)
+                {
+                    std::mdspan expectation_vals_j = std::submdspan(expectation_vals, j, std::full_extent);
+                    pauli_strings[j].expectation_value(expectation_vals_j, states);
+                }
 
-            // einsum("jk,jt->kt", coeffs, expectation_vals)
-#pragma omp parallel for collapse(2)
-            for (size_t t = 0; t < n_data; ++t)
-            {
+                // einsum("jk,jt->kt", coeffs, expectation_vals)
+                std::vector<std::complex<T>> coeffs_T_raw(n_ops * n_pauli_strings());
+                std::mdspan coeffs_T(coeffs_T_raw.data(), n_ops, n_pauli_strings());
+
+#pragma omp for schedule(static) collapse(2)
                 for (size_t k = 0; k < n_ops; ++k)
                 {
                     for (size_t j = 0; j < n_pauli_strings(); ++j)
                     {
-                        expectation_vals_out(k, t) += coeffs(j, k) * expectation_vals(j, t);
+                        coeffs_T(k, j) = coeffs(j, k);
+                    }
+                }
+
+#pragma omp for collapse(2) schedule(static)
+                for (size_t k = 0; k < n_ops; ++k)
+                {
+                    for (size_t t = 0; t < n_data; ++t)
+                    {
+                        for (size_t j = 0; j < n_pauli_strings(); ++j)
+                        {
+                            expectation_vals_out(k, t) += coeffs(j, k) * expectation_vals(j, t);
```

Comment:
Having `expectation_vals_out` shaped as (n_ops, n_times) kinda disagrees with exp val interface we established for PauliString and PauliOp where output has shape of (n_times,). I agree it might be useful to have it like that but I don't know if we currently have a use case for that?